underfs/s3/src/main/java/tachyon/underfs/s3/S3UnderFileSystem.java

```diff
@@ -0,0 +1,486 @@
+    try {
+      path = stripPrefixIfPresent(path);
+      path = path.endsWith(PATH_SEPARATOR) ? path : path + PATH_SEPARATOR;
+      path = path.equals(PATH_SEPARATOR) ? "" : path;
+      // Gets all the objects under the path, because we have no idea if there are non Tachyon
+      // managed "directories"
+      S3Object[] objs = mClient.listObjects(mBucketName, path, "");
+      if (recursive) {
+        String[] ret = new String[objs.length];
+        for (int i = 0; i < objs.length; i ++) {
+          // Remove parent portion of the key
+          String child = objs[i].getKey().substring(path.length());
+          // Prune the special folder suffix
+          child =
+              child.endsWith(FOLDER_SUFFIX) ? child.substring(0,
+                  child.length() - FOLDER_SUFFIX.length()) : child;
+          ret[i] = child;
+        }
+        return ret;
+      }
+      // Non recursive list
+      Set<String> children = new HashSet<String>();
+      for (S3Object obj : objs) {
+        // Remove parent portion of the key
+        String child = obj.getKey().substring(path.length());
+        // Remove any portion after the path delimiter
+        int childNameIndex = child.indexOf(PATH_SEPARATOR);
+        child = childNameIndex != -1 ? child.substring(0, childNameIndex) : child;
+        // Prune the special folder suffix
+        child =
+            child.endsWith(FOLDER_SUFFIX) ? child.substring(0,
+                child.length() - FOLDER_SUFFIX.length()) : child;
+        // Add to the set of children, the set will deduplicate.
+        children.add(child);
+      }
+      return children.toArray(new String[children.size()]);
+    } catch (ServiceException se) {
+      LOG.error("Failed to list path " + path);
+      return null;
+    }
+  }
+
+  /**
+   * Creates a directory flagged file with the key and folder suffix.
+   * @param key the key to create a folder
+   * @return true if the operation was successful, false otherwise
+   */
+  private boolean mkdirsInternal(String key) {
+    try {
+      String keyAsFolder = convertToFolderName(stripPrefixIfPresent(key));
+      S3Object obj = new S3Object(keyAsFolder);
+      obj.setDataInputStream(new ByteArrayInputStream(new byte[0]));
+      obj.setContentLength(0);
+      obj.setContentType(Mimetypes.MIMETYPE_BINARY_OCTET_STREAM);
+      mClient.putObject(mBucketName, obj);
+      return true;
+    } catch (ServiceException se) {
+      LOG.error("Failed to create directory: " + key, se);
+      return false;
+    }
+  }
+
+  /**
+   * Strips the s3 bucket prefix from the path if it is present. For example, for input key
+   * s3n://my-bucket-name/my-path/file, the output would be my-path/file.
+   * @param key the key to strip
+   * @return the key without the s3 bucket prefix
+   */
+  private String stripPrefixIfPresent(String key) {
+    String prefix = Constants.HEADER_S3N + mBucketName + PATH_SEPARATOR;
```

Comment:
NIT: this prefix really seems a constant after constructor is called, maybe worth have a private member string to prevent string calucation each time on stripPrefixIfPresent